Download SBOM in promotion pipeline

chmeliik opened the pull request:
This is a hacky way to integrate the SBOM-downloading into the promotion pipeline

* We need to call the gather-images script twice with different params. Work around that by adding another script that calls it with the params we need.
* The second run of gather-images will just overwrite the results of the first one. This is fine at the moment, because the entire pipeline runs serially (I think?). It would still be possible to run the two `gitops_steps` in parallel if they each ran in a separate environment.

chmeliik commented:
Tested with the local pipeline by adding this to `promote-pipeline.sh` (`IMAGE` is the result of `build-pipeline.sh`)

```diff
diff --git a/promote-pipeline.sh b/promote-pipeline.sh
index 9bbe8d1..a81cfcd 100755
--- a/promote-pipeline.sh
+++ b/promote-pipeline.sh
@@ -7,6 +7,11 @@ export LOCAL_SHELL_RUN=true
 
 source setup-local-dev-repos.sh
 
+for f in tmp/gitops/components/tssc-dev/overlays/*/deployment-patch.yaml; do
+    IMAGE="quay.io/acmiel-test/bootstrap:jenkins-7beb3bb72c9f7fe00cebe0d552c148f4cce3085b" \
+        yq '.spec.template.spec.containers[].image = strenv(IMAGE)' -i "$f"
+done
+
 REQUIRED_ENV="MY_QUAY_USER "
 REQUIRED_BINARY="tree "
 rhtap/verify-deps-exist "$REQUIRED_ENV" "$REQUIRED_BINARY"
```

and running `STRICT=false bash promote-pipeline.sh` (to ignore the failing EC)

chmeliik commented:
There's a small(?) problem: if the stage or prod image in the gitops repo are not built by the build pipeline and signed with the correct key, the pipeline will fail with something like this

```
Getting attestation for quay.io/redhat-appstudio/dance-bootstrap-app:latest
Failed to verify any attestation type. Errors:
  
  Command: cosign verify-attestation --type=slsaprovenance02 --key /tmp/download-sbom-workdir.lNson4/cosign.pub --insecure-ignore-tlog quay.io/redhat-appstudio/dance-bootstrap-app:latest
  
  WARNING: Skipping tlog verification is an insecure practice that lacks of transparency and auditability verification for the attestation.
  Error: no matching attestations: 
  main.go:74: error during command execution: no matching attestations: 
  
  Command: cosign verify-attestation --type=slsaprovenance1 --key /tmp/download-sbom-workdir.lNson4/cosign.pub --insecure-ignore-tlog quay.io/redhat-appstudio/dance-bootstrap-app:latest
  
  WARNING: Skipping tlog verification is an insecure practice that lacks of transparency and auditability verification for the attestation.
  Error: no matching attestations: 
  main.go:74: error during command execution: no matching attestations: 
```

In the Tekton pipeline, this isn't much of an issue because the pipeline will only try to EC-verify / download SBOMs for _changed_ images, so the default one is ignored. This probably isn't handled in the Jenkins pipeline

simonbaird reviewed (approved):
I didn't test it, but looks good to me.

Thanks for separating it up into nice atomic commits.

chmeliik commented:
I tried an alternative approach that wouldn't require adding a wrapper script that calls the actual script (it feels quite confusing to me). Something like

```yaml
  - name: Upload SBOM
    substeps:
      - name: gather-images-to-upload-sbom
        script_name: gather-deploy-images
        args_str: "stage prod"
      - download-sbom-from-url-in-attestation
```

But the templates got really ugly really quickly :disappointed: 
If you'd like to see it I can open another PR with the alternative approach, but personally I think it's not worth it. With better templating it might be the nicer solution, but I don't have the jinja/nunchucks experience to make it work

chmeliik commented:
> There's a small(?) problem: if the stage or prod image in the gitops repo are not built by the build pipeline and signed with the correct key, the pipeline will fail with something like this
> 
> ```
> Getting attestation for quay.io/redhat-appstudio/dance-bootstrap-app:latest
> Failed to verify any attestation type. Errors:
>   
>   Command: cosign verify-attestation --type=slsaprovenance02 --key /tmp/download-sbom-workdir.lNson4/cosign.pub --insecure-ignore-tlog quay.io/redhat-appstudio/dance-bootstrap-app:latest
>   
>   WARNING: Skipping tlog verification is an insecure practice that lacks of transparency and auditability verification for the attestation.
>   Error: no matching attestations: 
>   main.go:74: error during command execution: no matching attestations: 
>   
>   Command: cosign verify-attestation --type=slsaprovenance1 --key /tmp/download-sbom-workdir.lNson4/cosign.pub --insecure-ignore-tlog quay.io/redhat-appstudio/dance-bootstrap-app:latest
>   
>   WARNING: Skipping tlog verification is an insecure practice that lacks of transparency and auditability verification for the attestation.
>   Error: no matching attestations: 
>   main.go:74: error during command execution: no matching attestations: 
> ```
> 
> In the Tekton pipeline, this isn't much of an issue because the pipeline will only try to EC-verify / download SBOMs for _changed_ images, so the default one is ignored. This probably isn't handled in the Jenkins pipeline

What do you think about this? I don't think it's a new problem, I think it already affects EC verification in the promotion pipeline no?

cc @simonbaird @jduimovich

chmeliik commented:
Rebased on main

chmeliik commented:
> > > In the Tekton pipeline, this isn't much of an issue because the pipeline will only try to EC-verify / download SBOMs for _changed_ images, so the default one is ignored. This probably isn't handled in the Jenkins pipeline
> > 
> > 
> > What do you think about this? I don't think it's a new problem, I think it already affects EC verification in the promotion pipeline no?
> 
> There used to be code in the gather images which tested this case and removed the image as a target image. This code seems to have been dropped. It should be added back
> 
> ```
> 	  # Workaround for RHTAPBUGS-1284
> 	  if [[ "$image" =~ "quay.io/redhat-appstudio/dance-bootstrap-app" ]]; then
> 	    # Don't check the dance-bootstrap-app image
> 	    continue
> 	  fi
> ```

Hmm, I looked I through the git history here and in konflux-ci/build-definitions and didn't find that code. But it _does_ seem familiar. In Tekton, the workaround is not necessary because the detection of changed images works. As far as I can tell, it doesn't currently work in any other CI workflow (because nothing sets the TARGET_BRANCH for the gather-images script, and maybe because the clone is shallow and doesn't contain the ref anyway)

chmeliik commented:
Rebased on main, fixed formatting

chmeliik commented:
> Hmm, I looked I through the git history here and in konflux-ci/build-definitions and didn't find that code. But it _does_ seem familiar. In Tekton, the workaround is not necessary because the detection of changed images works. As far as I can tell, it doesn't currently work in any other CI workflow (because nothing sets the TARGET_BRANCH for the gather-images script, and maybe because the clone is shallow and doesn't contain the ref anyway)

I was playing with it a bit in Jenkins and didn't find any good solution. Setting `TARGET_BRANCH=main` just made the script ignore _all_ images, because the pipeline (at least the way I set it up) runs on the main branch, so obviously no images changed between `main` and `main`. I realized I don't know how to set up a Jenkins pipeline to run on PR.

TL;DR - I added that workaround for skipping dance-bootstrap-app. This should be ready for review

chmeliik commented:
(I also tested this together with #28 in Jenkins, so I'm fairly confident it works)

jduimovich commented:
> TL;DR - I added that workaround for skipping dance-bootstrap-app. This should be ready for review
Thanks, here is the code we had in 1.2 which must have fallen off the wagon

https://github.com/redhat-appstudio/tssc-sample-jenkins/blob/v1.2/resources/gather-deploy-images.sh#L29-L33

chmeliik commented:
> > TL;DR - I added that workaround for skipping dance-bootstrap-app. This should be ready for review
> > Thanks, here is the code we had in 1.2 which must have fallen off the wagon
> 
> https://github.com/redhat-appstudio/tssc-sample-jenkins/blob/v1.2/resources/gather-deploy-images.sh#L29-L33

Yeah I added it in the last commit

simonbaird commented on rhtap.groovy:
```diff
@@ -73,3 +73,11 @@ def gather_deploy_images( ) {
 }
+
+def gather_images_to_upload_sbom( ) {
+    run_script ('gather-images-to-upload-sbom.sh')
```
I you might need something similar to https://github.com/redhat-appstudio/tssc-dev-multi-ci/pull/43 to make sure the gather-deploy-images is present.

chmeliik replied:
Good point. Currently it works, but only because gather-deploy-images is also used as a standalone step before the gather-images-to-upload-sbom step. Will fix
